include/quda.h

```diff
@@ -1035,15 +1042,48 @@ extern "C" {
+   * sub-partitions: each sub-partition invert one or more rhs'.
+   * The QudaInvertParam object specifies how the solve should be performed on each sub-partition.
+   * Unlike @invertQuda, the interface also takes the host side gauge as input. The gauge pointer and
+   * gauge_param are used if for inv_param split_grid[0] * split_grid[1] * split_grid[2] * split_grid[3]
+   * is larger than 1, in which case gauge field is not required to be loaded beforehand; otherwise
+   * this interface would just work as @invertQuda, which requires gauge field to be loaded beforehand,
+   * and the gauge field pointer and gauge_param are not used.
+   * @param _hp_x       Array of solution spinor fields
+   * @param _hp_b       Array of source spinor fields
+   * @param param       Contains all metadata regarding host and device storage and solver parameters
+   * @param h_gauge     Base pointer to host gauge field (regardless of dimensionality)
+   * @param gauge_param Contains all metadata regarding host and device storage for gauge field
+   */
+  void invertMultiSrcQuda(void **_hp_x, void **_hp_b, QudaInvertParam *param, void *h_gauge, QudaGaugeParam *gauge_param);
+
+  /**
+   * @brief Really the same with @invertMultiSrcQuda but for staggered-style fermions, by accepting pointers
+   * to fat links and long links.
+   * @param _hp_x       Array of solution spinor fields
+   * @param _hp_b       Array of source spinor fields
+   * @param param       Contains all metadata regarding host and device storage and solver parameters
+   * @param milc_fatlinks     Base pointer to host **fat** gauge field (regardless of dimensionality)
+   * @param milc_longlinks    Base pointer to host **long** gauge field (regardless of dimensionality)
+   * @param gauge_param Contains all metadata regarding host and device storage for gauge field
    */
-  void invertMultiSrcQuda(void **_hp_x, void **_hp_b, QudaInvertParam *param);
+  void invertMultiSrcStaggeredQuda(void **_hp_x, void **_hp_b, QudaInvertParam *param, void *milc_fatlinks,
+                                   void *milc_longlinks, QudaGaugeParam *gauge_param);
+
+  /**
+   * @brief Really the same with @invertMultiSrcQuda but for clover-style fermions, by accepting pointers
+   * to direct and inverse clover field pointers.
+   * @param _hp_x       Array of solution spinor fields
+   * @param _hp_b       Array of source spinor fields
+   * @param param       Contains all metadata regarding host and device storage and solver parameters
+   * @param h_gauge     Base pointer to host gauge field (regardless of dimensionality)
+   * @param gauge_param Contains all metadata regarding host and device storage for gauge field
+   * @param h_clover    Base pointer to the direct clover field
+   * @param h_clovinv   Base pointer to the inverse clover field
+   */
+  void invertMultiSrcCloverQuda(void **_hp_x, void **_hp_b, QudaInvertParam *param, void *h_gauge,
```

Comment:
Can we unify this with `invertMultiSrcQuda`? Make `h_clover` and `h_clovinv` arguments to the full function, and then pass them in as `nullptr` for non-clover inversions?

I can imagine why we don't want to deal with the nightmare of merging this with staggered (for now).

Reply:
(I guess for `C` compatibility it'd be comparing with `0`/`NULL`, but whatever)

Reply:
In `lib/interface_quda.cpp` this is the indeed the case, but for interface I feel this is cleaner?

Reply:
Yeah, I saw that. I still think we should merge them because Clover is still a Wilson-type operator, but this won't be the hill I die on 😄